Applications/StableDiffusionExample/ContentView.swift

```diff
@@ -0,0 +1,310 @@
+                    GridRow {
+                        if evaluator.modelFactory.canUseNegativeText {
+                            TextField("negative prompt", text: $negativePrompt)
+                                .onSubmit(generate)
+                                .disabled(evaluator.progress != nil)
+                                #if os(visionOS)
+                                    .textFieldStyle(.roundedBorder)
+                                #endif
+                            Button(action: { prompt = "" }) {
+                                Label("clear", systemImage: "xmark.circle.fill").font(
+                                    .system(size: 10))
+                            }
+                            .labelStyle(.iconOnly)
+                            .buttonStyle(.plain)
+                        } else {
+                            EmptyView()
+                            EmptyView()
+                        }
+
+                        if evaluator.modelFactory.canShowProgress {
+                            Toggle("Show Progress", isOn: $showProgress)
+                        }
+                    }
+                }
+            }
+            .frame(minWidth: 300)
+        }
+        .padding()
+    }
+
+    private func generate() {
+        Task {
+            await evaluator.generate(
+                prompt: prompt, negativePrompt: negativePrompt, showProgress: showProgress)
+        }
+    }
+}
+
+/// Progress reporting with a title.
+struct Progress: Equatable {
+    let title: String
+    let current: Double
+    let limit: Double
+}
+
+/// Async model factory
+actor ModelFactory {
+
+    enum LoadState {
+        case idle
+        case loading(Task<ModelContainer<TextToImageGenerator>, Error>)
+        case loaded(ModelContainer<TextToImageGenerator>)
+    }
+
+    enum SDError: Error {
+        case unableToLoad
+    }
+
+    public nonisolated let configuration = StableDiffusionConfiguration.presetSDXLTurbo
+
+    /// if true we show UI that lets users see the intermediate steps
+    public nonisolated let canShowProgress: Bool
+
+    /// if true we show UI to give negative text
+    public nonisolated let canUseNegativeText: Bool
+
+    private var loadState = LoadState.idle
+    private var loadConfiguration = LoadConfiguration(float16: true, quantize: false)
+
+    public nonisolated let conserveMemory: Bool
+
+    init() {
+        let defaultParameters = configuration.defaultParameters()
+        self.canShowProgress = defaultParameters.steps > 4
+        self.canUseNegativeText = defaultParameters.cfgWeight > 1
+
+        // this will be true e.g. if the computer has 8G of memory or less
+        self.conserveMemory = MLX.GPU.memoryLimit < 8 * 1024 * 1024 * 1024
+
+        if conserveMemory {
+            print("conserving memory")
+            loadConfiguration.quantize = true
+            MLX.GPU.set(cacheLimit: 1 * 1024 * 1024)
+            MLX.GPU.set(memoryLimit: 3 * 1024 * 1024 * 1024)
```

Comment:
Set the limits low if conserving memory -- every little bit helps.  This also tells the configuration to apply quantization to the model as it is loaded:

```
            if loadConfiguration.quantize {
                quantize(model: sd.textEncoder, filter: { k, m in m is Linear })
                quantize(model: sd.textEncoder2, filter: { k, m in m is Linear })
                quantize(model: sd.unet, groupSize: 32, bits: 8)
            }
```